### ISSUE TITLE
Linux precompiled releases stopped

### ISSUE THREAD

**Issue by arrowgent** [ZDoom/Raze#160]
hello

since v0.6.0 there have been no pre-compiled linux releases
what happened?

https://github.com/coelckers/Raze/pull/44
https://github.com/coelckers/Raze/pull/60

these commit/pr dont seem to have been accepted
im not entirely sure which ones are related to dropping the precompiles

any details what can be done to get pre-compiled linux assets back on the releases?
i will do what i can to help

thanks

**Comment by arrowgent** [comment]
moving source compile below:

compiling source doesnt seem to work either, or am i doing something wrong (following gzdoom's wiki)
v0.7.3 source

```
- Looking for clock_gettime in rt - found
CMake Error at /usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
  Could NOT find ZMusic (missing: ZMUSIC_LIBRARIES ZMUSIC_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindZMusic.cmake:21 (find_package_handle_standard_args)
  source/CMakeLists.txt:416 (find_package)


-- Configuring incomplete, errors occurred!
```
hmmm... let me go try this
https://forum.zdoom.org/viewtopic.php?t=67477

```
[ 67%] Building CXX object source/CMakeFiles/raze.dir/core/screenjob.cpp.o
/home/catbox/Downloads/g_raze/Raze-0.7.3_alpha/source/core/screenjob.cpp:51:10: fatal error: vpx/vpx_decoder.h: No such file or directory
 #include <vpx/vpx_decoder.h>
          ^~~~~~~~~~~~~~~~~~~
compilation terminated.
make[2]: *** [source/CMakeFiles/raze.dir/build.make:1096: source/CMakeFiles/raze.dir/core/screenjob.cpp.o] Error 1

```


going to move on until refactoring the new codebase hits a new milestone.
best of luck

**Comment by alexey-lysiuk** [comment]
Install `libvpx` development package. In APT it’s called `libvpx-dev`.

**Comment by arrowgent** [comment]
thats good to know, your "how to compile" section pointing to gzdoom wiki should probally include a seperate Raze section

**Comment by arrowgent** [comment]
compiled everything updated.
tested blood, duke3d, shadow warrior

so... back to the OP, what happened with precompiled linux releases?

**Comment by sinisterseed** [comment]
Well, you could say nothing.

They were distributed by @madame-rachelle , but she stopped. Presumably she has her reasons.

**Comment by arrowgent** [comment]
i see.  i was not aware only one person was making the precompiled binaries and packaging them

is there an autobuild possiblitiy for the repo?
i know for .deb there is usually a considerable effort getting someone onboard

thanks for the insight, voluntary contributions and all.

**Comment by madame-rachelle** [comment]
I have tried asking @Blzut3 if he is willing to do them but he seems reluctant to do it. Yes, compiling .debs, at least without scripts to help you, is a pain in the rear. It's even worse since Raze requires zmusiclite and I don't have a .deb for that yet.

**Comment by alexey-lysiuk** [comment]
It’s possible to build static ZMusic library by setting `BUILD_SHARED_LIBS` CMake variable to `OFF`.
Executable needs to be linked with the following libraries from ZMusic build directory:

- source/libzmusiclite.a
- thirdparty/dumb/libdumb.a
- thirdparty/game-music-emu/gme/libgme.a

**Comment by madame-rachelle** [comment]
I will try and get this done later today. I can't do much for a few hours at least right now.

**Comment by mjr4077au** [comment]
Appreciate any effort you can spare, so thank you 😁🤘

**Comment by Blzut3** [comment]
I have too much queued up to do so haven't gotten around to it.  We're looking at December at the earliest that I can get around to new build stuff.

**Comment by mjr4077au** [comment]
Thanks mate, appreciate all the effort. If worse comes to worse, I could always do a build in a VM I suppose but I've sort of always had the mindset that if a user is opting to use Linux then they probably should be able to build code.

If we did a .deb file, then someone will ask for .rpm file for Fedora, then someone will say the .deb file targeting some Ubuntu version doesn't work with their bullshit niche Debian fork no one has ever heard of, etc. I just don't know where you'd stop with distributing Linux binaries.

**Comment by madame-rachelle** [comment]
We'd stop at .deb of course.

Debian is the most widely supported and used family of Linux distros without much of a close second, although I do see Arch gaining a lot of popularity, but as far as I know Arch does things a little bit differently, anyhow.

That being said if someone like Blzut3 were to step up and say ... "hey, I can compile this for my system and I use X and I want to help out" - I certainly wouldn't say no.

**Comment by mjr4077au** [comment]
Yeah that's a fair assessment. Hopefully the .deb is fairly universal and doesn't run into dependency hell with specific versions of libs needed etc. The game is fairly low key on external deps which is good.

I wouldn't bother with a package for Arch Linux, someone in the community has done a PKGBUILD in Arch's AUR which is awesome and works well. If it's in the AUR, an AUR helper like yay can be used to get updates from the command line as new releases are done (so long as the maintainer updates the PKGBUILD of course).

I wonder if the script that as geeing these [Linux devbuilds](https://devbuilds.drdteam.org/raze-linux/) up just needs a poke? Not sure what under the hood was building these and why it stopped.

**Comment by Blzut3** [comment]
Windows binaries might not work on that one person using Windows PE as a desktop operating system.  (I for one was sad that I couldn't run ZDoom on an HPe RAID controller's firmware image because it didn't have DirectX.)  It's easy to determine where to stop, it's where you think is reasonable.  The reason we ship debs is that I personally use Ubuntu and like to be able to just add a repo to apt and have the stable release version.  I do try to follow Canonical's support period so I build binaries for the oldest LTS and static link any non-ABI stable libraries.  Which for the most part means anyone using a distro released in the last 5 years can run the binaries by extract them from the deb.  Since the dependency list isn't anything crazy this works most of the time.

As an aside debs have the nice advantage of just being tar balls inside of an .ar archive so you don't need any special tools to extract the contents on non-deb distros.  Now of course some distros use different SONAMEs for the ABI stable libraries, and some distros might not be glibc based which can pose a problem.  But people using those distros probably aren't super interested in running binary blobs anyway.

GZDoom does have CPack rules, and CPack has gotten a lot better since those were added.  Ideally I'd like to modernize GZDoom and Raze's CMake projects so that I can switch to just using CPack to build the debs.  It would be trivial then to provide RPMs although the CentOS/Fedora ecosystem is in my opinion a little harder to support for various reasons, and this isn't the medium to discuss those details.  Personally though I don't use CentOS/Fedora so I don't have any intention of maintaining a yum repository.

No real comment on Arch, but given the way the AUR seems to work (by the way I don't use Arch) I can't imagine there's anything we'd really be able to distribute ourselves even if we wanted to (and as you may imagine, I'm not personally interested).

tl;dr supporting Linux binaries is really about finding that sweet spot where you cover the vast majority of users.  Distros are mostly compatible and even where they're not compatible out of the box some simple tricks can get the Ubuntu binaries running anyway.

**Comment by mjr4077au** [comment]
Agree with everything you've said there. Using CPack does sound interesting though, is there much work involved to do the uplift you've mentioned? (never looked into it before, sorry).

**Comment by Blzut3** [comment]
I wouldn't say it's necessarily a lot of work, but rewriting it confidently can be hard when you need to untangle a massive web of old-style CMake.  The easiest thing someone could do ahead of me finding time to do the conversion would be to start removing variables.  With modern CMake there's basically no reason to have variables that hold the names of source files or libraries to be linked.  You can just add an target with the unconditional files and then use target_sources to add all the platform specific stuff after the fact.  Call target_link_libraries, target_include_directories, etc as many times as needed.

Once everything is using modern targets and all the global variables are removed, zmusic can be added as an optional submodule which will allow CPack to include it in the package.

Outside of that, this would be a good excuse for me to update the GZDoom package so that it follows current packaging guidelines instead of the long since seemingly retracted Ubuntu 3rd party software guidelines.  Add a few formalities so that lintian checks can pass.

**Comment by mjr4077au** [comment]
@Blzut3 I hope you've been well, and Merry Christmas. Just wondering if you've had any chance to ponder updating the CMake stuff with CPack for GZDoom and when or if you might have the time to embark on this project?

**Comment by Blzut3** [comment]
I have a WIP for ZMusic on my laptop, will probably find some time to get that finished up this weekend if not next weekend.  Hopefully cleaning up GZDoom should follow quickly after that.

**Comment by mjr4077au** [comment]
Sounds great! 😁
